exploration/error-handling.md

```diff
@@ -0,0 +1,145 @@
+
+* Programming language: the language of the implementation informs idiomatic patterns of error handling.
+In Java, errors are thrown and subsequently caught in `try...catch` block.
+In Rust, fallible callsites (those which can return errors) should return a `Result<T, Err>` monad. 
+In both languages, built-in error handling assumes a singular error.
+* Environment constriants: as mentioned in [feedback from ICU4X](https://github.com/unicode-org/message-format-wg/issues/782#issuecomment-2103177417),
+ICU4X operates in low resource environments for which returning at most 1 error is desirable
+because returning more than 1 error would require heap allocation.
+* Programming conventions and idioms: in [feedback from ICU-TC](https://docs.google.com/document/d/11yJUWedBIpmq-YNSqqDfgUxcREmlvV0NskYganXkQHA/edit#bookmark=id.lx4ls9eelh99),
+they found over the 25 years of maintaining the library that there was more cost than benefit in additionally providing error information with a default best effort return value compared to just returning the default best effort value.
+The additional constraint in ICU4C's C++ style to return an error code rather than throwing errors using the STL further complicates the usefulness and likelihood to be used correctly by developers, especially during nested calls.
+
+> [!NOTE]
+> The wording in this document uses the word "signal" in regards to providing
+> information about an error rather than "return" or "emit" when referring to
+> a requirement that an implementation must at least indicate that an error has
+> occurred.
+> The word "signal" better accomodates more alternatives in the solution space
+> like those that only choose to indicate that an error occurred,
+> while still including those that additionally prefer to return the error
+> itself as an error object.
+> (By contrast, "return an error" implies that an error object will be thrown or
+> returned, and "emit an error" is ambiguous as to what is or isn't performed.)
+
+## Proposed Design
+
+The following spec text is proposed:
+
+> In all cases, when encountering an error,
+> a message formatter MUST be able to signal an error or errors.
+> It MAY also provide the appropriate fallback representation of the _message_ defined
+> in this specification.
+
+This solution requires implementations to be able to signal an error occurred,
+which can be accomplished in different ways. Ex:
+
+* an API that throws or returns an Error object when encountering an error
+* an API that includes a read/write `ErrorCode` argument 
+* two APIs, a permissive one that always returns a best-effort formatted result
+and a stricter one that throws or returns an Error object when encountering an error
+* an API that always returns a best-effort formatted result
+and a global `boolean` values
+
+This does not give implementations full freedom to return _nothing_ or some other behavior.
+
+## Alternatives Considered
```

Comment:
As discussed in https://github.com/unicode-org/message-format-wg/pull/804#issuecomment-2158131626:

```suggestion
## Alternatives Considered

### Separate Requirements

The following spec text is proposed:

> When formatting a message with one or more errors:
> - An implementation MUST provide a way for a user to be informed
>   of the name of at least one of the errors,
>   either directly or via an identifying error code.
> - An implementation MUST provide a way for a message with one or more
>   _Resolution Errors_ or _Message Function Errors_ to be formatted
>   using a fallback representation.
>
> The two above requirements MAY be fulfilled by a single formatting method,
> or separately by more than one such method.

This alternative ensures that even messages with runtime errors
can be formatted with the same string representation in all implementations.

Compared to the current spec text,
the under-specified "informative error" is replaced
with a more concrete minimum requirement,
which can be fulfilled with a two-byte error code in very limited environments.

The phrase "In all cases" is left out,
and the ability to fulfill the requirements separately
rather than at the same time is explicitly called out.
```